### ISSUE TITLE
V8JS php instable after a "die" or PHP error during a callback

### ISSUE THREAD

**Issue by sourisdudesert** [phpv8/v8js#282]
At first, thanks for your work on phpv8.

I successfully manage to use V8JS with PHP/Apache for my project, but I discover than V8JS engine seems to become instable when a "die" happens.

It seems non reproductible according to #122 , but I reproduce this 100% of the time.

**Symptom :**
V8JS works perfectly fine until the first "die" happens. Starting from this, all uses of V8JS by other following requests crashes randomly with no explanation/exception. The only possible way to get back to a stable situation is to restart Apache.

**Tested on 2 different systems :**
Linux Mint 17.2  / Ubuntu 16.04
Apache 2.4.7 / 2.4.18
V8JS 5.7 (on both system)
phpv8 1.3.3 (on both system, with all options / defaults)

**Crash test code used :**
[crash.php.txt](https://github.com/phpv8/v8js/files/695843/crash.php.txt)

**What this crash test file is doing :**
_ during the 5 first steps, the V8JS javascript script makes a callback to a PHP function that "die()".
_ during all the steps afterwards, the V8JS javascript makes a simple callback to a PHP function that returns a normal results.

**What happens :**
_ The behavior when PHP die is correct (the whole script die).
_ The behavior during the steps afterwards is not correct, as the V8JS crashes and the "Normal end" message do not appear ~50% of the time. The V8JS seems to be frozen.

**How to recover from this :**
restart Apache.

**Conclusion & questions :**
I know that using "die" is a bad practice, but this bug gives the feeling that something bad may happens to V8JS that crashes all requests afterwards. This prevents me to go with V8JS in production, because I cannot risk to crash the whole server if something is going wrong with a single request.

1. shouldn't phpv8 be totally "stateless", so that 1 request crash do not impact any other request?
2. is it safe to use phpv8 in a production environment at now?

Thanks again for your work & answer.

**Comment by stesie** [comment]
hey, thanks for your detailed report

I agree that phpv8 should be totally stateless (over requests) as you put it. Yet that's not exactly what's going on in your system (and it even cannot do otherwise). Regardless the SAPI you're using (fpm or apache mod_php) one php process handles several (yet with NTP SAPI only subsequent) requests ... and this php process has V8 loaded and initialized (once for all requests).

If your script dies then php internally does a so-called bailout, i.e. resets the scripts memory buffer and just jumps out of all stack frames. Yet this is a problem with V8 since that way it doesn't know about the script's crash. Hence phpv8 already has special handlers for that case (yet obviously not enough), actually I started out with that with pull request #91.

Regarding your question whether phpv8 is safe to use in a production, as you point out, obviously (and unfortunately) not to 100%. Contrary it depends on the use-case. On my last job (where I actually started to care for phpv8) it wasn't much of a problem as V8Js was only run in PHP CLI SAPI which doesn't do multiple requests in one process.

TL;DR thanks for poking, I'll have another look at this, ... yet it may take some time

**Comment by sourisdudesert** [comment]
Thanks for your answer. Phpv8 is really precious to me as I have to allow my users to write server side scripts on my server, and all alternatives are outdated (ex : PHP runkit) or un-maintained (ex : PHP with Lua).

2 additional questions if you don't mind :

_ Do you think I can consider that I long as I don't use "die" anywhere in PHP phpv8 is safe for use with apache/mod_php ?
_ Do you think this would be possible / a good idea to provide a way to reset entirely the phpv8 state ? For example using "clearPendingException" ?

Again, thanks for your work.

**Comment by stesie** [comment]
Well `die` is not the only thing that triggers bailout. Uncaught exceptions or PHP error just do the same.

Do you have just single pages that require V8Js? Or is it more like the whole application? If it's just the former you might want to use fpm for those requests (and configure it to run a very little number of requests per process only). Anyways the only viable way forward is to get this issue fixed :)

Regarding your question whether to reset phpv8 state on errors, that's not quite possible. After all the PHP process loads V8 engine and initializes it once (and shuts it down if the process goes down; not after every single request). The problem is that V8 engine does not allow no restart it, once it was shut down (within one process; but anyways, it would be a bad idea since it just costs time)

**Comment by stesie** [comment]
I think I've found what you're experiencing, but not perfectly sure as you said crashing above. Is it just that V8Js somewhat misbehaves or does it really crash the whole php process (or apache child in that case)? So do you see segfault notices in `dmesg` for example?

The bug I've found was, that V8Js doesn't clear the "fatal error" flag as needed.
So if V8Js exitted due to a die, fatal error, uncaught exception, etc. once, then any subsequent `executeString` would silently bail out unexpectedly.

The reason why you have to do several steps calling `die` is that you (probably) have multiple php worker processes and any single one serves one such die well.

So please try the fix on issue-282 branch from https://github.com/stesie/v8js/tree/issue-282

**Comment by sourisdudesert** [comment]
Hello,

This is awesome, thanks a lot.

I've added the extra line on v8js.cc, recompile everything and it worked like a charm (ie the requests with "die" are failing, but the subsequent requests without "die" are not).

More infos :
_ I didn't see any segfault notices in dmesg. Before the fix the apache childs were exiting silently, providing an empty request to the client. This was making the situation particularly difficult because there were no reliable way to even detect the issue.
_ The bug you describe with the fatal error flag is coherent with everything I experienced.
_ The fact the error was "random" is definitely linked to the fact there was several Apache workers. To reproduce the error with Apache the best thing to do is to decrease the number of worker.
_ About the context : I'm using V8JS+phpV8 to run a multiplayer online game on boardgamearena.com platform (id you want to try it : game "Hack Trick" is using the V8JS engine). I'd like to allow the players community to build new games on my platform, so I need to allow script execution on server side. I tested several alternatives and V8JS+phpV8 is definitively the best to execute user-written scripts on server side.

Thanks again for your help.
